src/commands/keygen.rs

```diff
@@ -0,0 +1,380 @@
+                } else {
+                    return Err("'.private' already exists".into());
+                }
+            }
+
+            if let Ok(metadata) = std::fs::symlink_metadata(".key") {
+                if self.force_symlinks {
+                    if metadata.is_symlink() {
+                        std::fs::remove_file(".key")
+                            .map_err(|err| format!("could not remove symlink '.key': {err}"))?;
+                    } else {
+                        return Err("'.key' already exists but is not a symlink".into());
+                    }
+                } else {
+                    return Err("'.key' already exists".into());
+                }
+            }
+
+            if digest_file.is_some() {
+                if let Ok(metadata) = std::fs::symlink_metadata(".ds") {
+                    if self.force_symlinks {
+                        if metadata.is_symlink() {
+                            std::fs::remove_file(".ds")
+                                .map_err(|err| format!("could not remove symlink '.ds': {err}"))?;
+                        } else {
+                            return Err("'.ds' already exists but is not a symlink".into());
+                        }
+                    } else {
+                        return Err("'.ds' already exists".into());
+                    }
+                }
+            }
+        }
+
+        // Prepare the contents to write.
+        let secret_key = secret_key.display_as_bind().to_string();
+        let public_key = public_key.display_as_bind().to_string();
+        let digest = digest.map(|digest| {
+            format!(
+                "{} IN DS {}\n",
+                self.name.fmt_with_dot(),
+                digest.display_zonefile(false)
+            )
+        });
+
+        // Write the key files.
+        secret_key_file
+            .write_all(secret_key.as_bytes())
+            .map_err(|err| {
+                format!("error while writing private key file '{base}.private': {err}")
+            })?;
+        public_key_file
+            .write_all(public_key.as_bytes())
+            .map_err(|err| format!("error while writing public key file '{base}.key': {err}"))?;
+        if let Some(digest_file) = digest_file.as_mut() {
+            digest_file
+                .write_all(digest.unwrap().as_bytes())
+                .map_err(|err| format!("error while writing digest file '{base}.ds': {err}"))?;
+        }
+
+        secret_key_file.sync_all().map_err(|err| {
+            format!("error while writing private key file '{base}.private': {err}")
+        })?;
+        public_key_file
+            .sync_all()
+            .map_err(|err| format!("error while writing public key file '{base}.key': {err}"))?;
+        if let Some(digest_file) = digest_file.as_mut() {
+            digest_file
+                .sync_all()
+                .map_err(|err| format!("error while writing digest file '{base}.ds': {err}"))?;
+        }
```

Comment:
Is this really necessary?

Reply:
Not _particularly_, but Linux's system call interface says that an I/O error encountered by write A might actually be reported when performing write B (e.g. if write A resulted in a cache being updated, but the underlying device couldn't be written to by the time write B came around).

IMO, scripts could rely on `keygen` successfully exiting to assume that a perfectly usable keypair is now available to the caller.  I want to avoid such scripts from making dangerous changes (e.g. re-signing a zone) if the keys actually turn out to be unavailable or corrupted.

Reply:
I think that's a threat model that doesn't really apply, but alright.